src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,900 @@
+/*
+ *    Copyright (c) 2023 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "time-synchronization-server.h"
+#include "TimeSyncManager.h"
+#include "time-synchronization-delegate.h"
+
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app-common/zap-generated/ids/Attributes.h>
+#include <app-common/zap-generated/ids/Clusters.h>
+#include <app/AttributeAccessInterface.h>
+#include <app/CommandHandler.h>
+#include <app/EventLogging.h>
+#include <app/server/Server.h>
+#include <app/util/attribute-storage.h>
+#include <lib/support/CodeUtils.h>
+#include <lib/support/SortUtils.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <platform/CHIPDeviceLayer.h>
+
+#include <app-common/zap-generated/cluster-enums.h>
+
+#include <system/SystemClock.h>
+
+using namespace chip;
+using namespace chip::app;
+using namespace chip::DeviceLayer;
+using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::TimeSynchronization;
+using namespace chip::app::Clusters::TimeSynchronization::Attributes;
+using chip::TimeSyncDataProvider;
+using chip::Protocols::InteractionModel::Status;
+
+// -----------------------------------------------------------------------------
+// Delegate Implementation
+
+using chip::app::Clusters::TimeSynchronization::Delegate;
+
+namespace {
+
+Delegate * gDelegate = nullptr;
+
+Delegate * GetDelegate()
+{
+    if (gDelegate == nullptr)
+    {
+        static TimeSyncManager dg;
+        gDelegate = &dg;
+    }
+    return gDelegate;
+}
+} // namespace
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace TimeSynchronization {
+
+void SetDefaultDelegate(Delegate * delegate)
+{
+    gDelegate = delegate;
+}
+
+Delegate * GetDefaultDelegate()
+{
+    return GetDelegate();
+}
+
+} // namespace TimeSynchronization
+} // namespace Clusters
+} // namespace app
+} // namespace chip
+
+constexpr uint64_t kChipEpochUsSinceUnixEpoch = (uint64_t) kChipEpochSecondsSinceUnixEpoch * chip::kMicrosecondsPerSecond;
+
+static bool ChipEpochToUnixEpochMicro(uint64_t chipEpochTime, uint64_t & unixEpochTime)
+{
+    unixEpochTime = chipEpochTime + kChipEpochUsSinceUnixEpoch;
```

Comment:
I don't see how casting helps.  The question is: if `chipEpochTime` here is something like 2^64-1, what happens?  Can that happen?

Reply:
What to do when it happens somewhat depends on the set of circumstances in which it can happen, right?

That said, this fuction has a boolean return value precisely so it can indicate failure, right?

I don't see how "casting" solves this problem in any way, but maybe I don't understand what the proposal is?

Reply:
Tracked in https://github.com/project-chip/connectedhomeip/issues/27575